room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspType.kt

```diff
@@ -0,0 +1,174 @@
+ *
+ * We don't necessarily have a [KSTypeReference] (e.g. if we are getting it from an element).
+ * Similarly, we may not be able to get a [KSType] (e.g. if it resolves to error).
+ */
+internal class KspType private constructor(
+    private val env: KspProcessingEnv,
+    private val ksTypeReference: KSTypeReference?,
+    resolved: KSType?
+) : XDeclaredType, XEquality {
+    constructor(env: KspProcessingEnv, ksTypeReference: KSTypeReference) : this(
+        env = env,
+        ksTypeReference = ksTypeReference,
+        resolved = null
+    )
+
+    constructor(env: KspProcessingEnv, resolved: KSType) : this(
+        env = env,
+        ksTypeReference = null,
+        resolved = resolved
+    )
+
+    internal val ksType by lazy {
+        resolved ?: ksTypeReference?.resolve()
+    }
+    override val rawType by lazy {
+        KspRawType(this)
+    }
+    override val typeArguments: List<XType> by lazy {
+        // prioritize type reference if it exists as it will have actaul types user picked.
+        val args = ksTypeReference?.element?.typeArguments
+            ?: resolved?.arguments
+            ?: emptyList()
+        args.map {
+            // TODO what if it.type is null?
+            env.wrap(it.type!!)
+        }
+    }
+    override val typeName: TypeName by lazy {
+        ksType?.typeName() ?: ksTypeReference?.typeName() ?: UNDEFINED
+    }
+    override val nullability by lazy {
+        when (ksType?.nullability) {
+            Nullability.NULLABLE -> XNullability.NULLABLE
+            Nullability.NOT_NULL -> XNullability.NONNULL
+            else -> XNullability.UNKNOWN
+        }
+    }
+
+    override fun asTypeElement(): XTypeElement {
+        TODO("elements are not implemented yet")
+    }
+
+    override fun isAssignableFrom(other: XType): Boolean {
+        check(other is KspType)
+        return ksType.isAssignableFromWithErrorWorkaround(other.ksType)
+    }
+
+    override fun isError(): Boolean {
+        return ksType == null || ksType!!.isError
+    }
+
+    override fun defaultValue(): String {
+        val type = ksType ?: return "null"
+        // NOTE: this does not match the java implementation though it is probably more correct for
+        // kotlin.
+        if (type.nullability == Nullability.NULLABLE) {
+            return "null"
+        }
+        val builtIns = env.resolver.builtIns
+        return when (type) {
+            builtIns.booleanType -> "false"
+            builtIns.byteType, builtIns.shortType, builtIns.intType, builtIns.longType, builtIns
+                .charType -> "0"
+            builtIns.floatType -> "0f"
+            builtIns.doubleType -> "0.0"
+            else -> "null"
+        }
+    }
+
+    override fun boxed(): XType {
+        return this
+    }
+
+    override fun isInt(): Boolean {
+        return ksType.isAssignableFromWithErrorWorkaround(env.resolver.builtIns.intType)
```

Comment:
good call. it is actually a mystery for us how we want to handle kotlin nullability wrt type equality.

I'm deferring those decisions for now until we go into end-to-end testing to understand use cases better.

in this particular case though, we seem to really want to check w/o equality.

Reply:
actually i realized that i have a test for this which was passing because i'm checking assignability on ksType <- int not int <- ksType.

latter is the right way so i'll add a cache into the KspProcessingEnvironment for nullable versions of builtins that i can reference here.

Reply:
Good to know, actually you can just do a type.makeNullable() call and the result is cached as well so you don't need to worry about performance. Alternatively we are also planning to add nullable builtin types..

Reply:
oh i added caching myself.
just to make sure i understand properly, where does it get cached when i call `builtins.int.makeNullable()`.
I would like to make sure repeatedly calling `buildins.int.makeNullable` won't cause a noticable performance lag unless we cache it.

Reply:
that being said, when ti comes to micro optimizations, we have a lot to go there (e.g. we are just using lazy instead of lazy publish :), so there is some unnecessary locking but i'm not putting much focus on that right now )